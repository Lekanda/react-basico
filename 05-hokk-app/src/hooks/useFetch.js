import { useState,useEffect, useRef } from 'react'

export const useFetch = (url) => {
    
    const isMounted = useRef(true);
    const [state, setState] = useState({loading:true, error:null, data:null, });

    useEffect(() => {
        return () => {
            isMounted.current = false;
        }

    }, [])


    useEffect(() => {

        setState({loading:true, error:null, data:null, });
        
        fetch(url)
            .then(response => response.json())
            .then(data => {

                    if(isMounted.current) {
                        setState({
                            loading:false,
                            error:null,
                            data 
                        })
                    }
            })
            .catch(error => setState({error, loading:false}));
            
    }, [url])

    return state;

}
