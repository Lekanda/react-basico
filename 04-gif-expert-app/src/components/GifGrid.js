import React from 'react'
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {

    const {data:images,loading} = useFetchGifs(category);


    return (
        <>
            <h3 className="titulo-cat animate__animated animate__fadeIn animate__slow">{category}</h3>

            { loading && <p className="loading animate__animated animate__flash animate__slow">Loading...</p> } 
            {/* //&&: Si es true hace el loading, si no nada */}

            <div className="card-grid contenedor">
                <ol>
                    {
                        images.map( img => (
                            <GifGridItem 
                                key={img.id} 
                                { ...img }
                            />
                        ))
                    }
                </ol>
            </div>
        </>
    )
}
// { ...img }: es una forma de pasar todos los atributos de un objeto a un componente. Se pasan los atributos individualmente.
// ({id,title,url}): Asi es como se recogen los datos en el componente GifGridItem.


GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

