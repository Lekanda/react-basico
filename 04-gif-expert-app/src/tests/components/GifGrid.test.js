import React from 'react';
import '@testing-library/jest-dom';
import {shallow} from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs'
jest.mock('../../hooks/useFetchGifs');


describe('Pruebas en <GifGrid />', () => {

    const category = 'One Punch';

    test('Comparar Snapshot', () => {
        
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
    })

    test('Debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquiercosa.jpg',
            title: 'Cualquier cosa'
        }];
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper).toMatchSnapshot();

    })
});