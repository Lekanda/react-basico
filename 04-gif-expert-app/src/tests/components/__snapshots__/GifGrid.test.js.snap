// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> Comparar Snapshot 1`] = `
<Fragment>
  <h3
    className="titulo-cat animate__animated animate__fadeIn animate__slow"
  >
    One Punch
  </h3>
  <p
    className="loading animate__animated animate__flash animate__slow"
  >
    Loading...
  </p>
  <div
    className="card-grid contenedor"
  >
    <ol />
  </div>
</Fragment>
`;

exports[`Pruebas en <GifGrid /> Debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="titulo-cat animate__animated animate__fadeIn animate__slow"
  >
    One Punch
  </h3>
  <div
    className="card-grid contenedor"
  >
    <ol>
      <GifGridItem
        id="ABC"
        key="ABC"
        title="Cualquier cosa"
        url="https://localhost/cualquiercosa.jpg"
      />
    </ol>
  </div>
</Fragment>
`;
